runtime/doc/api.txt

```diff
@@ -2684,6 +2684,17 @@ nvim_buf_set_extmark({buffer}, {ns_id}, {line}, {col}, {*opts})
                     • "blend": blend with background text color. Not supported
                       for "inline" virt_text.
 
+                  • lifetime : control the lifetime of the mark.
```

Comment:
I don't think lumping in `ephemeral` extmarks with a concept of extmark invalidation (as an active response to events) really makes sense. 

IMO I would call this option `"Invalidate"` or something like that, where `"line"` is the current only value (but leaves the door open to invalidate the extmark on other conditions in the future, like deleting the range of chars instead of the line). And then leave `ephemeral` be what it is.

Reply:
Fine by me, just went with @lewis6991's suggestion to be able to move forward with the sign refactor.

Reply:
I'm happy if we keep this separate from `ephemeral` and call it `invalidate`, however I would like us to align on what `ephemeral` really means and how it conceptually differs to `invalidate`.

Conceptually I understood ephemeral extmarks to be invalidated/removed at the end of the redraw cycle, which is why I suggested combining these together. I thought this would work out nicely since these options are mutually exclusive and thus reduces entropy and ambiguity.

I know internally they are implemented by skipping the marktree and are applied to the redraw state directly, but isn't that just an implementation detail?

Is there a more accurate description we can update `ephemeral` to so the difference between that and `invalidate` is more conceptually distinct?  I guess one significant difference is that `ephemeral` marks can't be inspected with `nvim_buf_get_extmarks`, though this could be treated as a caveat rather than a defining property.

For reference here's the current description:

```
                  • ephemeral : for use with |nvim_set_decoration_provider()|
                    callbacks. The mark will only be used for the current
                    redraw cycle, and not be permantently stored in the
                    buffer.
```

Reply:
Ok, I agree that decoration providers and ephemeral extmarks are underdocumented, both in concept and detail.

For reference, after having them and using them for some while, I think the refactor of the interface that _would_ make sense is more like this: (not saying it has to be done, but it would be more conceptually transparent). Deprecate `ephemeral=true` as suggested. But then do this: add a new lua-only function `vim.decoration_provider_emit_decoration()` . This is no longer an "extmark" which lives for any period of time, but just an event (or "yielded value" of a generator, etc)  which is fired in response in a lua event handler. "decoration" is no longer a synonym to "decorated extmark", but a valid thing on its own. (then we can speak of "ephemeral decorations" in contrast to "extmarks with decorations").

I'd be happy to expand more on this (and at least improve the documentation if not more) after some refactors to the internals I am currently working on.

>  I guess one significant difference is that ephemeral marks can't be inspected with nvim_buf_get_extmarks, though this could be treated as a caveat rather than a defining property.

It is a caveat that they cannot be logged/introspected in in-process tests _at all_, but it is a defining property that they cannot be inspected the same way. an added inspection interface will be more like inspecting an event log (not inspecting any "state").

Reply:
> Deprecate `ephemeral=true` as suggested. But then do this: add a new lua-only function `vim.decoration_provider_emit_decoration()`

Sounds good to me. But there is no need to deprecate `ephemeral=true` here then right? Can just be deprecated whenever this replacement `ephemeral decoration` interface is added.

Do we then go ahead with a boolean `invalidate` property? Or `invalidate == 'line'` or `invalidate == { 'line',  <???> }`?

Reply:
>  Can just be deprecated whenever this replacement ephemeral decoration interface is added.

Yes, the point is that is completely orthogonal to anything which happens in this PR.